clippy_lints/src/cargo/require_workspace_dependencies.rs

```diff
@@ -0,0 +1,66 @@
+use cargo_metadata::Metadata;
+use clippy_utils::diagnostics::span_lint;
+use if_chain::if_chain;
+use rustc_lint::LateContext;
+use rustc_span::source_map::DUMMY_SP;
+
+use super::REQUIRE_WORKSPACE_DEPENDENCIES;
+
+pub(super) fn check(cx: &LateContext<'_>, metadata: &Metadata) {
+    let manifest_path = &metadata.packages[0].manifest_path;
+
+    let Ok(manifest) = std::fs::read_to_string(manifest_path) else {
+        span_lint(
+            cx,
+            REQUIRE_WORKSPACE_DEPENDENCIES,
+            DUMMY_SP,
+            &format!("unable to read the crate manifest `{manifest_path}`"),
+        );
+        return;
+    };
+    let Ok(manifest) = toml::from_str::<Manifest>(&manifest) else {
```

Comment:
Well, this check could be easily implemented out of clippy with a bash script, and if there are significant obstacles to merge I'd rather reimplement this out of clippy.

Also, I do not understand what burden there is in parsing `Cargo.toml` the way I implement it. Maybe I am missing something, but even if we take into account all the potential version compatibility issues that may arise - it is easy to either implement support for a new manifest version if it is ever released for this niche case and if it is incompatible - this lint can simply be removed. Although maybe you see it as a death by a thousand cuts, and if you allow this you'd have to allow something else - I see how that would become a problem. Maybe it already is the situation - I don't know, and I would hope to see more elaboration on this, if you have time for it ofc :D It is probably boring to explain this all the time, but I'd just like to know where this restriction is rooted in.

I am sure many people would appreciate this functionality here and now instead of waiting for it to land properly after the involved process, so excuse me for not putting a pin on it just now and bothering you further with this. I am just hoping that maybe there is a way to skip the wait time and limit the amount of work required before I can use this.

Reply:
> Although maybe you see it as a death by a thousand cuts

Pretty much exactly this.

But let me explain myself a little further. And no need to apologize for asking :)

The tool/crate that defines the `Cargo.toml` format is `cargo`. So cargo could also change the format for example in an edition. Now Clippy would have to keep up and extend this lint with a check `if edition == X { parse_this_way() } else { parse_that_way() }`. Or if the format gets extended, we would have to extend this check in Clippy as well. But at the same time all this work was already done in the `cargo` project, and if it were exposed through the `metadata`, `cargo` would've updated the `metadata` as well and Clippy wouldn't need to touch the lint.

So the maintenance issue is two-fold: On the one hand, parsing the `Cargo.toml` ourselves means that Clippy and cargo always must stay in sync and do the same work twice (cross-project dependencies are a PITA). On the other hand, the Clippy code might get more and more complex over the years, making it harder to maintain.

So it might not be an issue now, but in the years to come it can grow bigger and bigger. That's why I'd like to avoid doing this in the first place.

You can of cause argue that in this case it is fine, because it's only a tiny part of the `Cargo.toml` file and it surely will never change. But even here I can't be sure when reviewing, if your implementation covers all possible workspace setups. Clippy maintainers don't know in-depth how `cargo` works.

> and if you allow this you'd have to allow something else

It's not so much that we "have to" allow something else. But more, that if we give in here and there it adds a tiny bit of maintenance burden here and there. And a temporary solution is the most permanent one. No one will ever clean up things like this (obviously exaggerated). One example of this is our [`clippy_utils/paths.rs`](https://github.com/rust-lang/rust-clippy/blob/master/clippy_utils/src/paths.rs) file: In theory we don't need that and can do everything with `diagnostic_items`. But the file rather grows in size than it shrinks. And this is because we allow such paths, because we did before.

> this lint can simply be removed

"We no longer want to maintain this lint" was never a reason to deprecate a lint before and I don't want it to become one. Deprecating a lint has also some process behind it that it has to go through (admittedly this is automated and not a big hassle nowadays).